core-primitives/stf-executor/src/lib.rs

```diff
@@ -0,0 +1,137 @@
+/*
+	Copyright 2021 Integritee AG and Supercomputing Systems AG
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(all(feature = "std", feature = "sgx"))]
+compile_error!("feature \"std\" and feature \"sgx\" cannot be enabled at the same time");
+
+#[cfg(all(not(feature = "std"), feature = "sgx"))]
+extern crate sgx_tstd as std;
+
+use ita_stf::hash::TrustedOperationOrHash;
+use itp_types::{OpaqueCall, H256};
+use std::vec::Vec;
+
+// re-export module to properly feature gate sgx and regular std environment
+#[cfg(all(not(feature = "std"), feature = "sgx"))]
+pub mod sgx_reexport_prelude {
+	pub use thiserror_sgx as thiserror;
+}
+
+pub mod error;
+pub mod traits;
+
+#[cfg(feature = "sgx")]
+pub mod executor;
+
+/// Hash results of an operation execution on the STF
+#[derive(Clone, Debug)]
+pub struct ExecutionHashes {
```

Comment:
Yes the naming is not ideal here. It's probably a combination of me not understanding exactly what it represents, and the thing having a confusing name before the refactoring.
I don't know why we have both a `OperationHash` and a `CallHash` - this is just ported over as it was before. I also don't exactly understand what the difference should be. `ExecutionHashes` should probably be replaced by a better name. Provided we still need `OperationHash` AND `CallHash`, what could the aggregation of those two be called?

Reply:
I think @haerdib introduced these (not sure), so she might correct my statements.

I have taken a look at the code. Deriving from the current usage of `handle_trusted_worker_call`, which is only at two places. The first time, we only care about the potential error, the second time, we are only interested in the `trusted_operation_hash`, which is put into a `call_hash` vector (!!!), I sense some code smell here. :)

I am confident, we can get rid of one. I actually think we should not use the trusted operations hash. Reasoning: this is an internal helper struct of the worker.  The client sends either `TrustedCall`s or `TrustedGetter`s AFAIK. If we want to give a block inclusion proof to a client in the future, it will be more straightforward if we hash the exact object the client has sent to the worker. The hash is not needed for anything else.

Reply:
I would simply use the following structure:

```
ExecutionResult {
     // hash of the executed trusted call.
    call_hash: H256
}
```

Reply:
I'd be happy to get rid of one of those hashes. And yes, we mix the names in the process, that adds to the confusion. We now use the `operation_hash` and use it in a `call_hashes` vector. And you suggest to only use the `call_hash` and get rid of the `operation_hash`? Won't that break something, because we're now returning a different hash value?

Reply:
I am almost certain that this hash is not checked anywhere yet. If the integration tests pass, we should be fine.

Reply:
Yes, you got my suggestion correctly.

Reply:
I think I introduced this operation hash because of the top pool. As it has been only handling trusted operations and was oblivious about calls/getters, to remove an operation, the operation hash was necessary. If we can remove it, that'd we great.

Reply:
In order to remove the operation from the top pool, we have a different path: we use the `top_or_hash` function, which in itself also computes the hash of an operation and returns it in a different struct.

Reply:
Ah only just now saw the comment of @clangenb:

> I think @haerdib introduced these (not sure), so she might correct my statements.
> 
> I have taken a look at the code. Deriving from the current usage of handle_trusted_worker_call, which is only at two places. The first time, we only care about the potential error, the second time, we are only interested in the trusted_operation_hash, which is put into a call_hash vector (!!!), I sense some code smell here. :)
> 
> I am confident, we can get rid of one. I actually think we should not use the trusted operations hash. Reasoning: this is an internal helper struct of the worker. The client sends either TrustedCalls or TrustedGetters AFAIK. If we want to give a block inclusion proof to a client in the future, it will be more straightforward if we hash the exact object the client has sent to the worker. The hash is not needed for anything else.
> 

Yes, I introduced it, but I don't remember exactly why. Anway: The client uses trusted operations when sending it directly to the enclave: https://github.com/integritee-network/worker/blob/d0f47ba3f3c7f78d1deb2d343b469a6aee7aed6b/cli/src/main.rs#L440-L443

BUt calls when the indirect way is used..

Reply:
Our CLI yes, but I think it matters more, what we send to the worker because this is what 3rd party tools would do.

Reply:
Ah, wait, it is wrapped again in a top, but I think this is very unnecessary.

Reply:
Since we're using the same rpc call for both getters (signed & unsigned) as well as for calls, it's not completely unnecessary. But since we're encoding anyway, I guess we could also try to decode it to either getter or call and only return error when failed. If we do that, we might as well completely remove the trusted operation - or what else is it used for?

Reply:
Hm I think I lost track of this discussion now.. I removed the trusted operation hash now in the result, using only the call hash (in the one case where we gather them and write them into a confirmation block). Is that okay for both of you?

Reply:
I think that's fine if everything is still working. But we might want to create an issue to discuss the use case of TrustedOperation itself @clangenb  ?

Reply:
Yes, I think @mullefel change is fine.

I also agree to discuess the usage of trusted op.

Reply:
https://github.com/integritee-network/worker/issues/489

Reply:
fyi: I had to revert back to using the operation hash (not the call hash), since we have tests that check for that. So to be on the safe side, I'll use the operation hash for now to not alter any behavior in the code.

Reply:
Ok, even better. 👍